pkg/cert-generator/v1beta1/generator.go

```diff
@@ -0,0 +1,280 @@
+)
+
+// CertGenerator is the manager to generate certs.
+type CertGenerator struct {
+	namespace      string
+	serviceName    string
+	controllerName string
+	kubeClient     client.Client
+	certsReady     chan struct{}
+
+	certs             *certificates
+	fullServiceDomain string
+}
+
+var _ manager.Runnable = &CertGenerator{}
+var _ manager.LeaderElectionRunnable = &CertGenerator{}
+
+func (c *CertGenerator) Start(ctx context.Context) error {
+	if err := c.generate(ctx); err != nil {
+		return err
+	}
+	// Close a certsReady means start to register controllers to the manager.
+	close(c.certsReady)
+	return nil
+}
+
+func (c *CertGenerator) NeedLeaderElection() bool {
+	return true
+}
+
+// AddToManager adds the cert-generator to the manager.
+func AddToManager(mgr manager.Manager, config configv1beta1.CertGeneratorConfig, certsReady chan struct{}) error {
+	return mgr.Add(&CertGenerator{
+		namespace:      consts.DefaultKatibNamespace,
+		serviceName:    config.ServiceName,
+		controllerName: config.ControllerName,
+		kubeClient:     mgr.GetClient(),
+		certsReady:     certsReady,
+	})
+}
+
+// generate generates certificates for the admission webhooks.
+func (c *CertGenerator) generate(ctx context.Context) error {
+	controllerService := &corev1.Service{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Namespace: c.namespace, Name: c.serviceName}, controllerService); err != nil {
+		return fmt.Errorf("%w: %v", errServiceNotFound, err)
+	}
+
+	certExist, err := c.isCertExist(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCertCheckFail, err)
+	}
+	if !certExist {
+		c.fullServiceDomain = strings.Join([]string{c.serviceName, c.namespace, "svc"}, ".")
+
+		if err = c.createCert(); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertFail, err)
+		}
+		if err = c.createCertSecret(ctx); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertSecretFail, err)
+		}
+	}
+	if err = c.saveCertOnLocal(); err != nil {
+		return fmt.Errorf("%w: %v", errSaveCertOnLocal, err)
+	}
+	if err = c.injectCert(ctx); err != nil {
+		return fmt.Errorf("%w: %v", errInjectCertError, err)
+	}
+	return nil
+}
+
+// isCertExist checks if a secret embedded certs already exists.
+// For example, it will return true if the katib-controller is created with enabled leader-election
+// since another controller pod will create the secret.
+func (c *CertGenerator) isCertExist(ctx context.Context) (bool, error) {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: Secret, Namespace: c.namespace}, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	key := secret.Data[serverKeyName]
+	cert := secret.Data[serverCertName]
+	if len(key) != 0 && len(cert) != 0 {
+		c.certs = &certificates{
+			keyPem:  key,
+			certPem: cert,
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// saveCertOnLocal saves the certs on local.
+func (c *CertGenerator) saveCertOnLocal() error {
```

Comment:
Should we verify that certs files exist ?

Reply:
Do you mean that so that we can use the local certs?

Reply:
I removed this function since we should use certs embedded on Secret instead of certs on local.

ref: https://github.com/kubeflow/katib/pull/2185#discussion_r1281835847